### PR TITLE
Fix AppBar to can have automatic Backbutton and endDrawer together

### LINKED ISSUE #97608: Automatic BackButton on AppBar disappears if there is an endDrawer open
see e.g. https://stackoverflow.com/questions/66485355/flutter-2-0-appbar-back-button-disappeared-if-contains-enddrawer

Looks like our fix for https://github.com/flutter/flutter/pull/63272 was a little too heavy-handed and we need a more nuanced approach.

### PR DESCRIPTION
**Pull request by jja08111**
This PR fixes #97608.

Related PRs: #63272, #80845

<details>
<summary>sample code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(home: MyApp()));
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      endDrawer: const Drawer(),
      onEndDrawerChanged: (_) {
        setState(() {});
      },
      floatingActionButton: FloatingActionButton(
        child: const Icon(Icons.add),
        onPressed: () {
          Navigator.push(
            context,
            MaterialPageRoute(
              builder: (_) => StatefulBuilder(builder: (context, setState) {
                return Scaffold(
                  appBar: AppBar(),
                  endDrawer: const Drawer(),
                  onEndDrawerChanged: (_) {
                    setState.call(() {});
                  },
                );
              }),
            ),
          );
        },
      ),
    );
  }
}
```

</details>

### Before

https://user-images.githubusercontent.com/57604817/154826705-73117abf-f0fa-4a7b-99a0-2e3bef9ee15a.mov

### After

https://user-images.githubusercontent.com/57604817/154826688-8a0601c3-0e8a-4274-8314-f73ae7fb358d.mov

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -881,8 +883,9 @@ class _AppBarState extends State<AppBar> {
           tooltip: MaterialLocalizations.of(context).openAppDrawerTooltip,
         );
       } else {
-        if (!hasEndDrawer && canPop)
+        if (canPop && !(hasEndDrawer && isFirstRoute && isEndDrawerOpen)) {
```
Comment (xeinebiu) [inline packages/flutter/lib/src/material/app_bar.dart:886]:
```suggestion
final ____ !(hasEndDrawer && isFirstRoute && isEndDrawerOpen);
if (canPop && ____) {
```
Just to make it more clear what is the final purpose of the `!(hasEndDrawer && isFirstRoute && isEndDrawerOpen)`. Replace the `____` with a meaningful variable name.

Comment (jja08111) [reply packages/flutter/lib/src/material/app_bar.dart]: @xeinebiu I've replaced it. And also changed the parameter name that wasn't fit the style guide in the test file.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 2]
```diff
@@ -881,8 +883,10 @@ class _AppBarState extends State<AppBar> {
           tooltip: MaterialLocalizations.of(context).openAppDrawerTooltip,
         );
       } else {
-        if (!hasEndDrawer && canPop)
+        final bool canHaveLeadingWithEndDrawer = !(hasEndDrawer && isFirstRoute && isEndDrawerOpen);
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/app_bar.dart:886]: if `isEndDrawerOpen` is true, the `hasEndDrawer` must also be true?

I think we can remove `hasEndDrawer`.

**Review by chunhtai — COMMENTED** [review]
overall looks good, left a minor comment

File: packages/flutter/lib/src/material/app_bar.dart (hunk 2 above)
Comment (jja08111) [reply packages/flutter/lib/src/material/app_bar.dart]: Thanks, good point. I removed the `hasEndDrawer`.

**Comment by chunhtai** [comment]
it looks like you may need to rebase off latest master branch

**Comment by jja08111** [comment]
Oh, there is too many changes after rebasing. I have to recreate PR.

**Pull request closed without merging**